### PR TITLE
Refactor unblinded tokens

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/16005

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Confirm tokens are refilled after successfully fetching issuers from `GET /v1/issuers`
- Confirm tokens are refilled after successfully redeeming a token
- Confirm tokens are refilled after failing to redeem a token redeeming a token
- Confirm users are rewarded for viewing ads
- Confirm tokens are cashed-out
- Confirm `creativeType` (i.e. `ad_notification`, `new_tab_page_ad`, `inline_content_ad` or `promoted_content_ad`) is not included in the payload when tokens are cashed-out
- Confirm issuers are rejected if any payment issuer has more than two public keys with the same value
- Confirm issuers are rejected if the payment issuer is missing from  `GET /v1/issuers/` endpoint
- Confirm issuers are rejected if issuers have not changed since the last update
- Confirm issuers are fetched from `GET /v1/issuers/` endpoint on browser launch
- Confirm issuers are fetched when ads are enabled
- Confirm issuers are not fetched when updating the catalog
- Confirm issuers are fetched every x milliseconds (see `ping` in `GET /v1/issuers/` response)
- Confirm "like" button (thumbs up) on 30-day ads history sends an upvote confirmation to the server
- Confirm "dislike" button (thumbs down) on 30-day ads history sends a downvote confirmation to the server
- Confirm "save" button on 30-day ads history sends a saved confirmation to the server
- Confirm "mark as inappropriate" button on 30-day ads history sends a flag confirmation to the server
- Confirm clicking an ad sends a click confirmation to the server
- Confirm converting an ad sends a conversion confirmation to the server
- Confirm landing an ad (transferred) sends landed a confirmation to the server
- Confirm `unblinded_payment_tokens` node in `confirmations.json` is migrated to include `ad_type` (empty string), `confirmation_type` (empty string) and `transaction_id` (random v4 UUID)
- Confirm viewed ads are shown within 30-day ads history
- Confirm issuers are persisted in `confirmations.json`
- Confirm `unblinded_payment_tokens` node in `confirmations.json` has `ad_type`, `confirmation_type` and `transaction_id` for new entries
- Confirm catalog is downloaded from `GET /v9/catalog` endpoint

NOTE: `Failed to get issuers` will be shown after upgrading because the `issuers` node will only be created after successfully downloading issuers from `GET /v1/issuers/` for the first time. See the spec for more detail

### CODE REVIEW

**Review by emerick — APPROVED** [review]
Rewards changes LGTM

**Review by kylehickinson — APPROVED** [review]
iOS lgtm

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 1]
```diff
@@ -1194,66 +1187,69 @@ void RewardsDOMHandler::OnGetAdsHistory(const base::ListValue& ads_history) {
...
     return;
   }
 
   AllowJavascript();
 
-  const std::string id = args->GetList()[0].GetString();
-  const std::string creative_set_id = args->GetList()[1].GetString();
-  const int action = args->GetList()[2].GetInt();
+  ads::AdContentInfo ad_content;
+  const base::Value value = args->GetList()[0].Clone();
```
Comment (aseren) [inline browser/ui/webui/brave_rewards_page_ui.cc:1199]: Clone() call is not needed here. We can use here a const reference:
`const base::Value& value = args->GetList()[0];`

or pass `args->GetList()[0]` directly to `ad_content.FromValue`

Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Fixed, thanks

**Comment by tmancey** [comment]
CI failed due to unrelated JUnit tests

**Review by aseren — APPROVED** [review]

**Review by bridiver — APPROVED** [review]
DEPS ok

**Comment by tmancey** [comment]
CI failing due to unrelated reasons (Storybook)

**Comment by tmancey** [comment]
CI failed for unrelated StoryBook issues

**Comment by tmancey** [comment]
CI failed during the upload of binary. Unrelated

**Pull request merged**